Sources/Testing/Running/Configuration.TestFilter.swift

```diff
@@ -422,3 +485,60 @@ extension Configuration.TestFilter {
     self = combining(with: other, using: op)
   }
 }
+
+// MARK: - Filterable types
+
+extension Configuration.TestFilter.Kind {
+  /// Whether this kind of test filter requires knowledge of test traits.
+  ///
+  /// If the value of this property is `true`, the values of a test graph must
+  /// be converted to `FilterItem` and have trait information recursively
+  /// propagated before the filter can be applied, or else the results may be
+  /// inaccurate. This facilitates a performance optimization where trait
+  /// propagation can be skipped for filters which don't require such knowledge.
+  fileprivate var requiresFilterItemConversion: Bool {
+    switch self {
+    case .unfiltered,
+         .testIDs,
+         .patterns:
+      false
+    case .tags:
+      true
+    case let .combination(lhs, rhs, _):
+      lhs.requiresFilterItemConversion || rhs.requiresFilterItemConversion
+    }
+  }
+}
+
+/// A protocol representing a value which can be filtered using
+/// ``Configuration/TestFilter-swift.struct``.
+private protocol _FilterableItem {
+  /// The test this item represents.
+  var test: Test { get }
+
+  /// The complete set of tags for ``test``, including those inherited from
+  /// containing suites.
+  var tags: Set<Tag> { get }
+}
+
+extension Test: _FilterableItem {
+  var test: Test {
+    self
+  }
+}
+
+/// An item representing a test and its filtering-related properties.
+///
+/// Instances of this type are needed when applying a test graph to a kind of
+/// filter for which the value of the `requiresFilterItemConversion` property
+/// is `true`.
+fileprivate struct FilterItem: _FilterableItem {
```

Comment:
How is this type fundamentally different from just using a test directly, other than caching `tags`?

Reply:
It doesn't just _cache_ tags, it collects all the tags for a test including those inherited from parents.